### PR TITLE
feat: permission stack traces in ops

### PR DESCRIPTION
**Pull request by crowlKats**
This commit improves permission prompts by adding an option
to print a full trace of where the permissions is being requested.

Due to big performance hint of stack trace collection, this is only
enabled when `DENO_TRACE_PERMISSIONS` env var is present.

Closes https://github.com/denoland/deno/issues/20756

### CODE REVIEW

File: cli/args/flags.rs [hunk 1]
```diff
@@ -3316,6 +3317,8 @@ fn permission_args(app: Command, requires: Option<&'static str>) -> Command {
                                              <p(245)>--deny-run  |  --deny-run="whoami,ps"</>
       <g>--deny-ffi[=<<PATH>...]</>               (Unstable) Deny loading dynamic libraries. Optionally specify denied directories or files.
                                              <p(245)>--deny-ffi  |  --deny-ffi="./libfoo.so"</>
+      <g>DENO_TRACE_PERMISSIONS</>               Environmental variable to enable stack traces in permission prompts.
+                                             <p(245)>DENO_TRACE_PERMISSIONS=1 deno run main.ts</>
```
Comment (crowlKats) [inline cli/args/flags.rs:3321]: added twice, once to the env docs and once to the perm docs for visibility

**Review by littledivy — APPROVED** [review]
LGTM!

File: runtime/permissions/prompter.rs [hunk 2]
```diff
@@ -354,6 +364,20 @@ impl PermissionPrompter for TtyPrompter {
...
+          writeln!(
+            &mut output,
+            "┃  {} {}",
+            colors::gray(if idx != len - 1 { "├─" } else { "└─" }),
+            colors::gray(deno_core::error::format_frame::<
+              deno_core::error::NoAnsiColors,
+            >(&frame))
+          )
+          .unwrap();
+        }
```
Comment (bartlomieju) [inline runtime/permissions/prompter.rs:379]: We need to account for the no of lines added here further down when we clear the terminal

File: runtime/permissions/prompter.rs [hunk 3]
```diff
@@ -354,6 +364,20 @@ impl PermissionPrompter for TtyPrompter {
...
+      }
```
Comment (bartlomieju) [inline runtime/permissions/prompter.rs:380]: Also if there's no stack trace we should print a line saying how it can be enabled

**Review by bartlomieju — CHANGES_REQUESTED** [review]

File: cli/args/flags.rs [hunk 4]
```diff
@@ -1157,25 +1157,26 @@ static ENV_VARIABLES_HELP: &str = cstr!(
...
+                          <p(245)>(defaults to "mozilla")</>
+  <g>HTTP_PROXY</>             Proxy address for HTTP requests
+                          <p(245)>(module downloads, fetch)</>
+  <g>HTTPS_PROXY</>            Proxy address for HTTPS requests
+                          <p(245)>(module downloads, fetch)</>
+  <g>NO_COLOR</>               Set to disable color
+  <g>NO_PROXY</>               Comma-separated list of hosts which do not use a proxy
+                          <p(245)>(module downloads, fetch)</>
+  <g>NPM_CONFIG_REGISTRY</>    URL to use for the npm registry.
+  <g>DENO_TRACE_PERMISSIONS</> Environmental variable to enable stack traces in permission prompts."#
```
Comment (bartlomieju) [inline cli/args/flags.rs:1179]: Maybe put it next to other `DENO_` env vars?

File: cli/worker.rs [hunk 5]
```diff
@@ -613,6 +613,9 @@ impl CliMainWorkerFactory {
       origin_storage_dir,
       stdio,
       skip_op_registration: shared.options.skip_op_registration,
+      enable_stack_trace_arg_in_ops: crate::args::has_flag_env_var(
+        "DENO_TRACE_PERMISSIONS",
+      ),
```
Comment (bartlomieju) [inline cli/worker.rs:618]: Maybe we could move it somewhere to `CliOptions` so it's only read once?

File: runtime/permissions/prompter.rs (hunk 3 above)
Comment (crowlKats) [reply runtime/permissions/prompter.rs]: done, however find it a bit odd now that it mentions the env var yet the handling of the env var is done in cli and  not permissions, but thats minor

**Review by bartlomieju — APPROVED** [review]
LGTM too

**Pull request merged**